scripts/build_test_history.py

```diff
@@ -0,0 +1,447 @@
+
+    # Since aligned runs need to all be completed runs to be fetched,
+    # a time window buffer of 24 hours is kept to allow runs to finish before
+    # assuming we've processed all aligned runs up to present time.
+    # Therefore, we only process runs up to (now - 24 hours).
+    yesterday = datetime.now() - timedelta(days=1)
+    end_interval = date_start_obj + timedelta(days=1)
+    if end_interval > yesterday:
+        end_interval = yesterday
+
+    end_interval_string = end_interval.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
+    url = (f'{RUNS_API_URL}?label=master'
+           '&label=experimental&max-count=1&aligned'
+           f'&from={date_start}&to={end_interval_string}')
+
+    try:
+        resp = requests.get(url)
+    # Sometimes this request can time out. If it does, just return
+    # an empty list and attempt the fetch again.
+    except requests.exceptions.ReadTimeout:
+        return []
+    runs_list: list[MetadataDict] = resp.json()
+
+    # If we have no runs to process in this date interval,
+    # we can skip this interval for processing from now on.
+    if len(runs_list) == 0:
+        print('No runs found for this interval.')
+        update_recent_processed_date(date_entity, end_interval_string)
+        # If we've processed up to (now - 24 hours), then return null,
+        # which signals we're done processing.
+        if end_interval == yesterday:
+            return None
+
+    # Sort by revision -> then time start, so that the aligned runs are
+    # processed in groups with each other.
+    # Note that this technically doesn't have an impact if only 1 set of
+    # aligned runs are processed, but this sort will allow this script to
+    # function properly if multiple aligned run sets were to be processed
+    # together.
+    runs_list.sort(key=lambda run: run['revision'])
+    runs_list.sort(key=lambda run: run['time_start'])
+
+    # Print the dates just to get info on the list of runs we're working with.
+    print('Runs to process:')
+    for run in runs_list:
+        print(f'ID: {run["id"]}, {run["browser_name"]} {run["time_start"]}')
+
+    return runs_list
+
+
+def update_recent_processed_date(
+        date_entity: MostRecentHistoryProcessed, new_date: str) -> None:
+    """Update the most recently processed date after finishing processing."""
+    date_entity.Date = new_date
+    date_entity.put()
+
+
+class NoRecentDateError(Exception):
+    """Exception raised when the MostRecentHistoryProcessed
+    entity is not found.
+    """
+    pass
+
+
+class JSONGenerationError(Exception):
+    """Exception raised when initial JSON files are being generated,
+    but the database has not been cleared of existing entries.
+    """
+    pass
+
+
+def get_processing_start_date() -> MostRecentHistoryProcessed:
+    most_recent_processed: MostRecentHistoryProcessed = (
+        MostRecentHistoryProcessed.query().get())
+
+    if most_recent_processed is None:
```

Comment:
Does it mean that it will raise an error when user runs it the first time?

Reply:
This will raise an error if there's no date entity in datastore that tells us what date to start the processing on. We created the entity manually in the datastore UI to start, but that doesn't seem intuitive. 😅  I've now I also added a flag that can be set to write this entity or set it to a different date 🙂 

```sh
# Set history processing start date to the beginning of 2023
python scripts/build_test_history.py --set-history-start-date=2023-01-01T00:00:00.000Z
```